### ISSUE TITLE
linux, segfault with a clang-built binary

### ISSUE THREAD

**Issue by sezero** [Novum/vkQuake#324]
This is an old outdated linux setup (fedora 27) with old drivers, I
don't expect much, but here it is:

Build vkquake using gcc-7.3.1 or gcc-9.2.1: no issues (other than
the annoying flicker, which should be something to do with my old
drivers, etc.)

Build vkquake using clang-5.0.2 or clang-11.1.0, it segfaults when
I try to start or load a game.  If I change `-O2` to `-Og` or remove
`-O2` from CFLAGS, no segfaults. Couldn't manage getting a meaningful
backtrace. Possibly a driver bug?
```
Vulkan Initialization
Using Vulkan 1.1
Using VK_KHR_surface
Using VK_KHR_xlib_surface
Using VK_KHR_get_surface_capabilities2
Using VK_KHR_get_physical_device_properties2
Vendor: Intel
Device: Intel(R) HD Graphics 630 (Kaby Lake GT2)
Using VK_KHR_swapchain
Using VK_KHR_get_memory_requirements2
Using VK_KHR_dedicated_allocation
Using A2B10G10R10 color buffer format
Using D24_S8 depth buffer format
Creating command buffers
Initializing staging
Creating descriptor set layouts
Reallocating dynamic VBs (256 KB)
Reallocating dynamic IBs (1024 KB)
Reallocating dynamic UBs (256 KB)
Initializing samplers
Creating pipeline layouts
Using IMMEDIATE present mode
Creating color buffer
AA disabled
Creating depth buffer
Creating render passes
Creating frame buffers
Creating pipelines

Thread 1 "vkquake" received signal SIGSEGV, Segmentation fault.
0x00007ffff67a38a3 in __memmove_avx_unaligned_erms () from /lib64/libc.so.6
(gdb) bt
#0  0x00007ffff67a38a3 in __memmove_avx_unaligned_erms () from /lib64/libc.so.6
#1  0x0000000000000021 in ?? ()
#2  0x0000000001dc55c0 in ?? ()
#3  0x0000000001eacfe0 in ?? ()
#4  0xee2feec5ffffffff in ?? ()
#5  0x00007fffffffdcf0 in ?? ()
#6  0x0000000000000000 in ?? ()
```

**Comment by Novum** [comment]
Not a driver bug, but without a backtrace it's impossible to debug without a repro.

**Comment by Novum** [comment]
Okay, I booted up my Linux and I can reproduce this. So that's good.

**Comment by sezero** [comment]
> Okay, I booted up my Linux and I can reproduce this. So that's good.

Great!

I'm sure you can do a better job than me at tracing it.

**Comment by sezero** [comment]
Out of curiosity: IIR MSVC can use clang as a backend, yes? (Never ever did
that myself, don't know how to set that up.)  Any issues with such a build?

**Comment by Novum** [comment]
Well, now I can't get it anymore. And the only thing that changed is that I ran the usual Ubuntu updates. What OS are you running?

**Comment by sezero** [comment]
>  What OS are you running?

An old fedora-27 (very outdated, I know.)

**Comment by Novum** [comment]
It seemstoo good to be true that the update fixed it, but it seems like it did. I can't reproduce at all anymore and valgrind is now also not complaining. Super strange.

**Comment by sezero** [comment]
Maybe it's something to with mesa backend (don't know how)?
What package versions changed with the update?

**Comment by Novum** [comment]
I'm running the proprietary NVIDIA driver. It installed a libc update, that would be my bet.

**Comment by sezero** [comment]
OK, the problem seems to be in CL_LoadCSProgs(), here: https://github.com/Novum/vkQuake/blob/master/Quake/host.c#L737
That `versionedname` there is _not_ initialized. If I specifically out
its use, no crashes. (valgrind still emits lots of warnings.)  Spike's
tree does have initialization for it:
https://github.com/Shpoike/Quakespasm/blob/master/quakespasm/Quake/host.c#L749

Without the initialization, PR_LoadProgs (https://github.com/Novum/vkQuake/blob/master/Quake/pr_edict.c#L1197)
gets a `com_filesize` of `-1` and mdfour makes a mess, hence the crash.

The ultra-magically curious thing is, how `COM_LoadMallocFile` returns
non-NULL result _and_ `-1` for com_filesize.

**Comment by sezero** [comment]
The solution to this bug would be sync'ing `CL_LoadCSProgs` to Spike's version.
(How that procedure was stripped down this much, or how other compilers/systems
didn't explode, I don't know.)

**Comment by sezero** [comment]
I pushed https://github.com/Novum/vkQuake/commit/4690bb85e95c5a2fc2bcd61c1b6969f7a89588cc as a _minimal_ fix to the immediate problem.

However: `CL_LoadCSProgs` was added by `https://github.com/Shpoike/Quakespasm/commit/7f9d54c63fe2397459c6951f1e931384492d1882
in Spike's tree, and several other functionality from there is missing
here. I'm not interested in those, but not closing this yet, in case you
want to remember those stuff.
